Bluetooth: controller: Consideration for SDU interval in fragmentation

niag-Oticon opened the pull request:
- Included consideration for the packet sequence number and time stamps together with the SDU interval in matching the event for fragmentation of SDUs to unframed PDUs.

- Included aunit test to cover change.

kruithofa commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1496,27 +1492,123 @@ static isoal_status_t isoal_tx_try_emit_pdu(struct isoal_source *source,
+ * @param  payload_number[out] Next payload number
+ * @return                     Number of skipped SDUs
+ */
+uint16_t isoal_tx_unframed_get_next_playload_number(isoal_source_handle_t source_hdl,
```
```suggestion
uint16_t isoal_tx_unframed_get_next_payload_number(isoal_source_handle_t source_hdl,
```

niag-Oticon replied:
Thanks. Amazing that that got through all the reviews.

niag-Oticon replied:
Done.

kruithofa commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1496,27 +1492,123 @@ static isoal_status_t isoal_tx_try_emit_pdu(struct isoal_source *source,
+					time_diff_valid && time_diff > session->sdu_interval) {
+				/* SDU time stamps are not consecutive if more
+				 * than two SDU intervals apart. Find the number
+				 * of skipped SDUs based of the difference in
```
```suggestion
				 * of skipped SDUs based on the difference in
```

niag-Oticon replied:
Done.

niag-Oticon committed:
Bluetooth: controller: Consideration for SDU interval in fragmentation

Included consideration for the packet sequence number and time stamps
together with the SDU interval in matching the event for fragmentation
of SDUs to unframed PDUs.

Implemented:
-- Increase in kconfig limit of BT_CTLR_ISO_TX_BUFFERS to allow full 32
   buffers to be available for LL/CIS/PER/BV-39C
-- ISO-AL microsecond time wrapping exposed to be used externally to
   adjust incoming Group Reference Points and Time-stamps where required
-- Function that handles adjustment of payload number according to
   incoming SDU exposed for external use to allow next payload number to
   be accessed by ISO transmit tests
-- Changed internal fragmentation source identification parameter from
   the source structure pointer to the source handle so that only the
   source handle needs to be used for external calls

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

carlescufi commented:
@cvinayak please review

cvinayak reviewed (changes requested):
Requesting change until investigation of why there is loss of audio sync

Thalley commented:
@niag-Oticon Please rebase

niag-Oticon commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -5717,27 +5718,33 @@ int hci_iso_handle(struct net_buf *buf, struct net_buf **evt)
+		sdu_frag_tx.target_event = cis->lll.event_count + event_offset;
+		sdu_frag_tx.grp_ref_point = isoal_get_wrapped_time_us(cig->cig_ref_point,
+						(event_offset * cig->iso_interval *
+							ISO_INT_UNIT_US));
```
Just thinking out aloud here:
I'm wondering if the source generating the audio is drifting with respect to the CIG reference point over time?
Since the CIG reference point is synchronized with the central it will most likely drift with respect to the periodic events of the source generating the SDUs. This would mean that even if the SDUs arrived in time to be fragmented in to payloads that could be transmitted when streaming started, as the CIG reference point drifts, the SDUs might start arriving later and finally too late for the sub-events in the target event.

niag-Oticon replied:
Any progress or updates?

cvinayak replied:
I still see loss of ISO data, I have BabbleSim tests for Multiple Central CIS support in draft PR that fail with change in this PR (it could be the changes in the draft PR too, but the tests pass without the changes in this PR).

The draft PR with the tests rebase to changes in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/56074

niag-Oticon replied:
The purpose of these changes are to ensure that we don't skew the stream of data when fragmenting SDUs. This is done by considering the SDU interval as well as the incoming SDU packet SN and time-stamp.

I debugged a bit on the PR that you mentioned. I see a couple of problems:

1. When all 8 connections are established, there appears to be a point where there is a lag in processing the channel requests for the 7th and 8th connections. As a result the request arriving at the ISO-AL appears to be delayed by more than an ISO interval or even more.

So for example if we consider before this PR and after for ISO Interval 10ms, BN=1, FT=1 :

**Before:**
T=12ms   Central       HCI:  SDU 1  ----- ISO-AL: Target Event 2 -----> PDU 2
T=20ms   Peripheral RX:    PDU 2  ----- ISO-AL: --------------------> SDU 2
T=22ms   Central       HCI:  SDU 2  ----- ISO-AL: Target Event 3 -----> PDU 3
T=30ms   Peripheral RX:    PDU 3  ----- ISO-AL: --------------------> SDU 3
T=32ms   Central       HCI:  Delayed
T=40ms   Peripheral RX:    FLUSH  ----- ISO-AL: --------------------> SDU 4 (No data Err) **
T=42ms   Central       HCI:  SDU 3  ----- ISO-AL: Target Event 5 -----> PDU 5
T=50ms   Peripheral RX:    PDU 5  ----- ISO-AL: --------------------> SDU 5 (Skewed, should have been SDU 4)
T=52ms   Central       HCI:  SDU 4  ----- ISO-AL: Target Event 6 -----> PDU 6
T=60ms   Peripheral RX:    PDU 6  ----- ISO-AL: --------------------> SDU 6 (Skewed, should have been SDU 5)

**After:**
T=12ms   Central       HCI:  SDU 1  ----- ISO-AL: Target Event 2 -----> PDU 2
T=20ms   Peripheral RX:    PDU 2  ----- ISO-AL: --------------------> SDU 2
T=22ms   Central       HCI:  SDU 2  ----- ISO-AL: Target Event 3 -----> PDU 3
T=30ms   Peripheral RX:    PDU 3  ----- ISO-AL: --------------------> SDU 3
T=32ms   Central       HCI:  Delayed
T=40ms   Peripheral RX:    FLUSH  ----- ISO-AL: --------------------> SDU 4 (No data Err)
T=42ms   Central       HCI:  SDU 3  ----- ISO-AL: Target Event 5 -----> PDU 3 [No skew, but too late for TX so flushed]
T=50ms   Peripheral RX:    FLUSH  ----- ISO-AL: --------------------> SDU 5 (No data Err)
T=52ms   Central       HCI:  SDU 4  ----- ISO-AL: Target Event 6 -----> PDU 4 [No skew, but too late for TX so flushed]
T=60ms   Peripheral RX:    FLUSH  ----- ISO-AL: --------------------> SDU 6 (No data Err)

2.  On the peripheral side, if only null PDUs are received, I don't see the error status of the PDUs sent to ISO-AL being set. This seems to be transformed in to a Start Continue PDU with zero length and no error status so it is interpreted as a padding PDU by the ISO-AL and no SDU is released. This means that in the previous example, the SDUs with status "No data Err" would not be received by the application.

niag-Oticon replied:
The "No Data Err" in the previous comment is actually "Lost Data" `ISOAL_PDU_STATUS_LOST_DATA`.

Thalley commented on tests/bluetooth/ctrl_isoal/src/sub_sets/isoal_test_tx.c:
```diff
@@ -2908,6 +2908,323 @@ ZTEST(test_tx_unframed, test_tx_unframed_1_sdu_1_frag_pdu_emit_err)
+	err = isoal_tx_get_sync_info(source_hdl, &tx_sync_seq, &tx_sync_timestamp, &tx_sync_offset);
+
+	zassert_equal(err, ISOAL_STATUS_OK, "err = 0x%02x", err);
+	zassert_equal(tx_sync_seq, tx_sync_seq_expected, "%d != %d", tx_sync_seq, 2);
+	zassert_equal(tx_sync_timestamp, tx_sync_timestamp_expected, "%d != %d", tx_sync_seq, 2);
+	zassert_equal(tx_sync_offset, tx_sync_offset_expected, "%ld != %ld", tx_sync_seq, 0);
```
Ditto

```suggestion
	zassert_equal(tx_sync_seq, tx_sync_seq_expected, "%u != %u", 
		      tx_sync_seq, tx_sync_seq_expected);
	zassert_equal(tx_sync_timestamp, tx_sync_timestamp_expected, "%u != %u", 
		      tx_sync_timestamp, tx_sync_timestamp_expected);
	zassert_equal(tx_sync_offset, tx_sync_offset_expected, "%u != %u", 
		      tx_sync_offset, tx_sync_offset_expected);
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/sub_sets/isoal_test_tx.c:
```diff
@@ -2908,6 +2908,323 @@ ZTEST(test_tx_unframed, test_tx_unframed_1_sdu_1_frag_pdu_emit_err)
+	testdata_size += 10;
+	sdu_fragments = 0;
+
+	isoal_test_create_sdu_fagment(BT_ISO_SINGLE,
+				&testdata[testdata_indx], (testdata_size - testdata_indx),
+				sdu_total_size, sdu_packet_number,
+				sdu_timestamp, ref_point, event_number,
+				&tx_sdu_frag_buf.sdu_tx);
```
Ditto

Thalley commented on tests/bluetooth/ctrl_isoal/src/sub_sets/isoal_test_tx.c:
```diff
@@ -2908,6 +2908,323 @@ ZTEST(test_tx_unframed, test_tx_unframed_1_sdu_1_frag_pdu_emit_err)
+	err = isoal_tx_get_sync_info(source_hdl, &tx_sync_seq, &tx_sync_timestamp, &tx_sync_offset);
+
+	zassert_equal(err, ISOAL_STATUS_OK, "err = 0x%02x", err);
+	zassert_equal(tx_sync_seq, tx_sync_seq_expected, "%d != %d", tx_sync_seq, 2);
+	zassert_equal(tx_sync_timestamp, tx_sync_timestamp_expected, "%d != %d", tx_sync_seq, 2);
+	zassert_equal(tx_sync_offset, tx_sync_offset_expected, "%ld != %ld", tx_sync_seq, 0);
```
Ditto

Thalley commented on tests/bluetooth/ctrl_isoal/src/sub_sets/isoal_test_tx.c:
```diff
@@ -2908,6 +2908,323 @@ ZTEST(test_tx_unframed, test_tx_unframed_1_sdu_1_frag_pdu_emit_err)
+	testdata_size += 10;
+	sdu_fragments = 0;
+
+	isoal_test_create_sdu_fagment(BT_ISO_SINGLE,
+				&testdata[testdata_indx], (testdata_size - testdata_indx),
+				sdu_total_size, sdu_packet_number,
+				sdu_timestamp, ref_point, event_number,
+				&tx_sdu_frag_buf.sdu_tx);
```
Ditto

Thalley commented on tests/bluetooth/ctrl_isoal/src/sub_sets/isoal_test_tx.c:
```diff
@@ -2908,6 +2908,323 @@ ZTEST(test_tx_unframed, test_tx_unframed_1_sdu_1_frag_pdu_emit_err)
+	err = isoal_tx_get_sync_info(source_hdl, &tx_sync_seq, &tx_sync_timestamp, &tx_sync_offset);
+
+	zassert_equal(err, ISOAL_STATUS_OK, "err = 0x%02x", err);
+	zassert_equal(tx_sync_seq, tx_sync_seq_expected, "%d != %d", tx_sync_seq, 2);
+	zassert_equal(tx_sync_timestamp, tx_sync_timestamp_expected, "%d != %d", tx_sync_seq, 2);
+	zassert_equal(tx_sync_offset, tx_sync_offset_expected, "%ld != %ld", tx_sync_seq, 0);
```
Ditto

niag-Oticon committed:
tests: bluetooth: controller: Unit test for SDU interval consideration

-- Unit test to check that the correct payload number is selected based
   on the SDU packet number and the time stamp when fragmenting SDUs
   into PDUs.
-- Ran clang-format on isoal_test_tx and isoal_test_rx

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

niag-Oticon commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -220,7 +220,7 @@ config BT_CTLR_ISO_RX_BUFFERS
 	int "Number of Isochronous Tx buffers"
 	depends on BT_CTLR_ADV_ISO || BT_CTLR_CONN_ISO
-	range 1 19
+	range 1 32
```
We need 31 at present so that we can send all fragments in a burst, the previous maximum was 19 so we increased it to accommodate what we require. If the maximum is higher, I don't know why it was set at 19.

Thalley replied:
Should we not just set the maximum to 0xFF now then?

niag-Oticon replied:
Done.

niag-Oticon commented on tests/bluetooth/ctrl_isoal/src/sub_sets/isoal_test_tx.c:
```diff
@@ -2908,6 +2908,323 @@ ZTEST(test_tx_unframed, test_tx_unframed_1_sdu_1_frag_pdu_emit_err)
+	sdu_read_loc = 0;
+	pdu_write_size = 23;
+
+	source_hdl = basic_tx_test_setup(0xADAD,           /* Handle */
+					 role,             /* Role */
+					 false,            /* Framed */
```
Because those aren't values that would have to change between test cases. The handle is not relevant the the tests so it only requires a place holder value and the tests for framed and unframed have different templates, so it's not possible to convert a framed test case to unframed or vice versa.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -5718,43 +5719,34 @@ int hci_iso_handle(struct net_buf *buf, struct net_buf **evt)
+		 * we can assume that we have enough time to deliver in the next
+		 * ISO event. If we're not active within the ISO event, we don't
+		 * know if there is enough time to deliver in the next event,
+		 * and for safety we set the target to current event + 2.
```
Is it a timing issue? 

i.e. we know we can complete it in the timespan on an event, but since we don't know how close we are to the end of the current event, we set the end of the next event as the deadline?

niag-Oticon replied:
Yes. If we target current event +1 we might be too close to the start of the event and the fragmented PDUs might not be ready in time for transmission.

Thalley replied:
OK, thanks, now I understand it :)

cvinayak commented:
@niag-Oticon please rebase

mtpr-ot reviewed (approved):
LGTM. EBQ passing with these changes on vendor platform.

cvinayak reviewed (approved):
NPI implementation in nRF LLL was incorrect, fixing it now has the CIS tests passing with the changes in this PR. Tests present in 0c3579b4a62a6f21b5543a60cad032d4a82d6ae7

niag-Oticon commented:
> Requesting change until investigation of why there is loss of audio sync

@cvinayak Thanks for the approval, but I believe that this change request is still blocking.

cvinayak commented:
> > Requesting change until investigation of why there is loss of audio sync
> 
> @cvinayak Thanks for the approval, but I believe that this change request is still blocking.

@kruithofa please review and approve.
